pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/commonarea"
+	"antrea.io/antrea/pkg/antctl/raw/multicluster/common"
+	multiclusterscheme "antrea.io/antrea/pkg/antctl/raw/multicluster/scheme"
+)
+
+type clusterOptions struct {
+	Namespace  string `yaml:"namespace,omitempty"`
+	ClusterID  string `yaml:"clusterID,omitempty"`
+	Secret     string `yaml:"secret,omitempty"`
+	Server     string `yaml:"server,omitempty"`
+	ClusterSet string `yaml:"clusterSet,omitempty"`
+}
+
+type joinOptions struct {
+	LeaderClusterOpt clusterOptions `yaml:"leaderCluster"`
+	MemberClusterOpt clusterOptions `yaml:"memberCluster"`
+	ConfigFile       string
+}
+
+type meta struct {
+	Kind       string `yaml:"kind"`
+	ApiVersion string `yaml:"apiVersion"`
+}
+
+var joinOpts *joinOptions
+
+func (o *joinOptions) validateAndComplete() error {
+	if o.ConfigFile != "" {
+		raw, err := ioutil.ReadFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		typeMeta := &meta{}
+		if err := yamlUnmarshall(raw, typeMeta); err != nil {
+			return err
+		}
+		if typeMeta.Kind != "ClusterSetConfig" || typeMeta.ApiVersion != "multicluster.antrea.io/v1alpha1" {
+			return fmt.Errorf("unknown apiVersion or kind in config: %s", o.ConfigFile)
+		}
+		if err := yamlUnmarshall(raw, &joinOpts); err != nil {
+			return err
+		}
+	}
+
+	if o.LeaderClusterOpt.ClusterID == "" {
+		return fmt.Errorf("the leader clusterID cannot be empty")
+	}
+	if o.LeaderClusterOpt.Server == "" {
+		return fmt.Errorf("the leader server cannot be empty")
+	}
+	if o.LeaderClusterOpt.Secret == "" {
```

Comment:
How the Secret is created in member? I think we should instead asking for a token string, and create a Secret (with a generated name, e.g. <clusterset>+"token") with the token string.

Reply:
I've thought about this way to set up. But I found that the token is so long to use it in the command line, and the Secret is an original kubernetes resources. So I think maybe the member can use `kubectl apply` to create Secret with the token distributed by the leader cluster.

Reply:
How long it is? I still believe we should print it and use the token to join. Later we can improve to add an option to save the token to file if it is really long, and we should add a command to get the token too.

Note, we add cli just o hide CDR/implementation details. It is very bad experience, we still require user to use kubectl to get and create resources. How can they remember that?

Reply:
I passed the token as an argument of the command, but got a x509 error. 
```
 antctl mc join --leader-namespace=antrea-multicluster \                                             1 ✘    with root@ubuntu  at 06:21:20  
--leader-clusterid=test-cluster-north \
    --leader-server=https://172.18.0.3:6443 \
    --clusterset=test-clusterset \
    --namespace=kube-system \
    --clusterid=test-cluster-east \
    --token=eyJhbGciOiJSUzI1NiIsImtpZCI6IkVLZGI0RTFRQXpGRDFTTXlSYjA4WUJaQ2M4aDRWeTVfakdVQU0yR1dwXzAifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6InRlc3Qtc2VjcmV0LTIiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoidGVzdC1zYS0yIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZXJ2aWNlLWFjY291bnQudWlkIjoiNDBjMTFkZTgtMGM1ZS00OTE3LTg5MWYtNDYyZGQwMGYyNmVkIiwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmRlZmF1bHQ6dGVzdC1zYS0yIn0.UpK4ekUKO1kBI1wj1lXr87YLio3GCL7GsLB2MKO3It4zqK34hsu6DMp2_ZzKvP4uqDlaq3xXX4jEjC1r0KgLrB4mskaCLTg8CrexOhdxvGwrQf-wG-GVfo0VYIendUY3YVGTGs9Z-_4GPPht9Wz3iAI8mdtiFuIgo0kZxB5FauqYhhG6KgEw20xhyU3eeT5QF7nhTmRhIG4OS5K6VGxhSYC7aKr5MgVfOFSWXB7UhXEkhUJpiZ-WQR1DgUwCENe7_smuyfT0-ytB_7ISipfqrfgOsw9ku7XpKon9KOu3HGi06CaGHrpOkfVDzwCYszCorvKJULseW4H8mqYQM1shGg \
    --kubeconfig /root/.kube/east

ClusterClaim "id.k8s.io" created in Namespace kube-system
ClusterClaim "clusterset.k8s.io" created in Namespace kube-system
ClusterSet "test-clusterset" created in Namespace kube-system
Waiting for member cluster ready
Waiting for ClusterSet ready
Failed to wait for member cluster ready: Get "https://172.18.0.3:6443/api?timeout=32s": x509: certificate signed by unknown authority
Failed to join the ClusterSet. Deleting the created resources.
ClusterClaim "id.k8s.io" deleted
ClusterClaim "clusterset.k8s.io" deleted
ClusterSet "test-clusterset" deleted
Error: Get "https://172.18.0.3:6443/api?timeout=32s": x509: certificate signed by unknown authority
```

Reply:
The output of `antctl mc create memberaccess` is 
```bash
You can now join any number of clusters by running the "antctl mc join" command on each member cluster with the following token
token:
eyJhbGciOiJSUzI1NiIsImtpZCI6IkVLZGI0RTFRQXpGRDFTTXlSYjA4WUJaQ2M4aDRWeTVfakdVQU0yR1dwXzAifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6InRlc3Qtc2VjcmV0LTIiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoidGVzdC1zYS0yIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZXJ2aWNlLWFjY291bnQudWlkIjoiNDBjMTFkZTgtMGM1ZS00OTE3LTg5MWYtNDYyZGQwMGYyNmVkIiwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmRlZmF1bHQ6dGVzdC1zYS0yIn0.UpK4ekUKO1kBI1wj1lXr87YLio3GCL7GsLB2MKO3It4zqK34hsu6DMp2_ZzKvP4uqDlaq3xXX4jEjC1r0KgLrB4mskaCLTg8CrexOhdxvGwrQf-wG-GVfo0VYIendUY3YVGTGs9Z-_4GPPht9Wz3iAI8mdtiFuIgo0kZxB5FauqYhhG6KgEw20xhyU3eeT5QF7nhTmRhIG4OS5K6VGxhSYC7aKr5MgVfOFSWXB7UhXEkhUJpiZ-WQR1DgUwCENe7_smuyfT0-ytB_7ISipfqrfgOsw9ku7XpKon9KOu3HGi06CaGHrpOkfVDzwCYszCorvKJULseW4H8mqYQM1shGg
```

Maybe we can store the ca and the token to a file (even a Secret YAML for the leader to distribute to the members?)

Reply:
How about when creating an access token. The command will store the YAML file of the Secret. And then when a member cluster joins the ClusterSet, they need to specify the Secret or path to the Secret YAML. If a YAML file is specified, the command will create the Secret. @jianjuns @luolanzone

Reply:
Ok. It works for me to save the Secret to a YAML file. Maybe we can add an optional argument for "init" and "create membertoken" like "--token-secret-file" to save the Secret. Later, we should also add a "get" command to save the default or a specified token to a file, and a "delete" command to delete a SA/Secret.

@luolanzone : please share your opinions.

Reply:
@jianjuns, I discussed this with @hjiajing  yesterday, adding an option to save the secret works for me, and we may also print out or save the sample config file for the `init` including the path of new secret, so user can use the file directly without writing a new one.

Reply:
Generating a config file by `init` sounds good to me. In that case, I feel better to include the whole token in the file, to avoid two files. But then for `join`, we should let the user-provided token override the default token in the config file. Maybe we can include a Secret YAML in the config file, and add an optional `--token-secret-file`  argument, which can override the default token (though I know it will be different from other arguments which cannot override config file).

However, as the PR has been there for long, I suggest to implement config file generating in another PR, together with other enhancements we talked about, inc. commands for get/delete tokens.